README.md
# Rake-Android

[![Build Status](https://travis-ci.org/skpdi/rake-android.svg?branch=develop)](https://travis-ci.org/skpdi/rake-android) [![Coverage Status](https://coveralls.io/repos/skpdi/rake-android/badge.svg?branch=develop&service=github)](https://coveralls.io/github/skpdi/rake-android?branch=develop) 

*[Change Log](https://github.com/skpdi/rake-android/blob/master/CHANGELOG.md)*

<br/>

*Rake* 는 안드로이드 앱에서 JSON 로그를 전송할 수 있도록 도와주는 라이브러리입니다.
  
- **no extra dependency** required
- supports from **android 4.0** (API Level 14)
- **small** (*~100K*)

자세한 내용은 위키를 참조해주세요.

- [Rake Android: Getting Started](https://github.com/skpdi/rake-document/wiki/1.-Rake-Android-(%ED%95%9C%EA%B5%AD%EC%96%B4))
- [Rake Android: App Example](https://github.com/skpdi/rake-android/tree/master/rake_example)
- [Rake Android: API](http://skpdi.github.io/rake-android/docs/0.4.7/com/rake/android/rkmetrics/RakeAPI.html)

## Development

### Build

```gradle
> METRIC_TOKEN_LIVE={TOKEN} METRIC_TOKEN_DEV={TOKEN} ./gradlew clean build
```

### Android Studio on OSX

Android Studio 사용시, gradle 이 환경변수를 읽지 못하므로 다음과 같이 터미널에 세팅한 후 IDE 실행
 
```
> launchctl setenv METRIC_TOKEN_LIVE {TOKEN}  
> launchctl setenv METRIC_TOKEN_DEV $METRIC_TOKEN_DEV {TOKEN} 
```

추가적으로 테스트 실행을 위해 Android Studio 에서 `run configuration` 환경변수에 다음을 등록

- `METRIC_TOKEN_LIVE`
- `METRIC_TOKEN_ENV`

## License

**Apache 2.0** 

- [Mixpanel: Android](https://github.com/mixpanel/mixpanel-android/blob/master/LICENSE)
- [JSnappy (for prior to Rake 0.3)](https://code.google.com/p/jsnappy/source/browse/trunk/LICENCE.txt)



CHANGELOG.md
# Change Log

## [0.6.1](https://github.com/skpdi/rake-android/tree/0.6.1) (2018-08-14)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.6.0...0.6.1)

## [0.6.0](https://github.com/skpdi/rake-android/tree/0.6.0) (2018-07-26)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.5.1...0.6.0)

## [0.5.1](https://github.com/skpdi/rake-android/tree/0.5.1) (2018-05-04)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.5.0...0.5.1)

## [0.5.0](https://github.com/skpdi/rake-android/tree/0.5.0) (2018-01-22)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.8...0.5.0)

## [0.4.8](https://github.com/skpdi/rake-android/tree/0.4.8) (2017-07-20)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.7...0.4.8)

## [0.4.7](https://github.com/skpdi/rake-android/tree/0.4.7) (2017-05-31)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.6...0.4.7)

## [0.4.6](https://github.com/skpdi/rake-android/tree/0.4.6) (2017-03-03)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.5...0.4.6)

## [0.4.5](https://github.com/skpdi/rake-android/tree/0.4.5) (2017-02-16)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.4...0.4.5)

## [0.4.4](https://github.com/skpdi/rake-android/tree/0.4.4) (2016-07-26)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.3...0.4.4)

## [0.4.3](https://github.com/skpdi/rake-android/tree/0.4.3) (2016-06-15)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.2...0.4.3)

**Implemented enhancements:**

- Remove flush metric [\#4](https://github.com/skpdi/rake-android/pull/4) ([1ambda](https://github.com/1ambda))

## [0.4.2](https://github.com/skpdi/rake-android/tree/0.4.2) (2016-03-10)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.1...0.4.2)

## [0.4.1](https://github.com/skpdi/rake-android/tree/0.4.1) (2016-01-11)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.4.0...0.4.1)

**Fixed bugs:**

- fix\(rake\): Preserve empty headers [\#2](https://github.com/skpdi/rake-android/pull/2) ([1ambda](https://github.com/1ambda))
- fix typo. [\#1](https://github.com/skpdi/rake-android/pull/1) ([NohSeho](https://github.com/NohSeho))

**Merged pull requests:**

- test\(shuttle\): Validate header, body RAKE-390 [\#3](https://github.com/skpdi/rake-android/pull/3) ([1ambda](https://github.com/1ambda))

## [0.4.0](https://github.com/skpdi/rake-android/tree/0.4.0) (2015-12-07)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.3.19...0.4.0)

## [0.3.19](https://github.com/skpdi/rake-android/tree/0.3.19) (2015-09-11)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.3.18...0.3.19)

## [0.3.18](https://github.com/skpdi/rake-android/tree/0.3.18) (2015-08-27)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.3.17...0.3.18)

## [0.3.17](https://github.com/skpdi/rake-android/tree/0.3.17) (2015-08-03)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.3.16...0.3.17)

## [0.3.16](https://github.com/skpdi/rake-android/tree/0.3.16) (2015-05-26)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.3.15...0.3.16)

## [0.3.15](https://github.com/skpdi/rake-android/tree/0.3.15) (2015-04-24)
[Full Changelog](https://github.com/skpdi/rake-android/compare/0.3.10...0.3.15)

## [0.3.10](https://github.com/skpdi/rake-android/tree/0.3.10) (2014-06-10)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


rake-iOS-Library/Rake/Base64.m
//
//  Base64.m
//  rake-iOS-Library
//
//  Created by Matt Gallagher on 2009/06/03.
//  Modified by JungHyun Kim on 2014/04/10
//  Copyright 2009 Matt Gallagher. All rights reserved.
//
//  Permission is given to use this source code file, free of charge, in any
//  project, commercial or otherwise, entirely at your risk, with the condition
//  that any redistribution (in part or whole) of source code must retain
//  this copyright and permission notice. Attribution in compiled projects is
//  appreciated but not required.
//
#if ! __has_feature(objc_arc)
#error This file must be compiled with ARC. Either turn on ARC for the project or use -fobjc-arc flag on this file.
#endif

#import "Base64.h"

//
// Mapping from 6 bit pattern to ASCII character.
//
static unsigned char base64EncodeLookup[65] =
"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

//
// Definition for "masked-out" areas of the base64DecodeLookup mapping
//
#define xx 65

//
// Mapping from ASCII character to 6 bit pattern.
//
static unsigned char base64DecodeLookup[256] =
{
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, 62, xx, xx, xx, 63,
    52, 53, 54, 55, 56, 57, 58, 59, 60, 61, xx, xx, xx, xx, xx, xx,
    xx,  0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14,
    15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, xx, xx, xx, xx, xx,
    xx, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40,
    41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
    xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx, xx,
};

//
// Fundamental sizes of the binary and base64 encode/decode units in bytes
//
#define BINARY_UNIT_SIZE 3
#define BASE64_UNIT_SIZE 4

//
// NewBase64Decode
//
// Decodes the base64 ASCII string in the inputBuffer to a newly malloced
// output buffer.
//
//  inputBuffer - the source ASCII string for the decode
//	length - the length of the string or -1 (to specify strlen should be used)
//	outputLength - if not-NULL, on output will contain the decoded length
//
// returns the decoded buffer. Must be free'd by caller. Length is given by
//	outputLength.
//
void *RK_NewBase64Decode(
                         const char *inputBuffer,
                         size_t length,
                         size_t *outputLength)
{
    if (length == 0) {
        length = strlen(inputBuffer);
    }
    
    size_t outputBufferSize = (length / BASE64_UNIT_SIZE) * BINARY_UNIT_SIZE;
    unsigned char *outputBuffer = (unsigned char *)malloc(outputBufferSize);
    
    size_t i = 0;
    size_t j = 0;
    while (i < length)
    {
        //
        // Accumulate 4 valid characters (ignore everything else)
        //
        unsigned char accumulated[BASE64_UNIT_SIZE] = {0, 0, 0, 0};
        size_t accumulateIndex = 0;
        while (i < length)
        {
            unsigned char decode = base64DecodeLookup[inputBuffer[i++]];
            if (decode != xx) {
                accumulated[accumulateIndex] = decode;
                accumulateIndex++;
                
                if (accumulateIndex == BASE64_UNIT_SIZE) {
                    break;
                }
            }
        }
        
        //
        // Store the 6 bits from each of the 4 characters as 3 bytes
        //
        outputBuffer[j] = (unsigned char)(accumulated[0] << 2) | (accumulated[1] >> 4);
        outputBuffer[j + 1] = (unsigned char)(accumulated[1] << 4) | (accumulated[2] >> 2);
        outputBuffer[j + 2] = (unsigned char)(accumulated[2] << 6) | accumulated[3];
        j += accumulateIndex - 1;
    }
    
    if (outputLength) {
        *outputLength = j;
    }
    return outputBuffer;
}

//
// NewBase64Decode
//
// Encodes the arbitrary data in the inputBuffer as base64 into a newly malloced
// output buffer.
//
//  inputBuffer - the source data for the encode
//	length - the length of the input in bytes
//  separateLines - if zero, no CR/LF characters will be added. Otherwise
//		a CR/LF pair will be added every 64 encoded chars.
//	outputLength - if not-NULL, on output will contain the encoded length
//		(not including terminating 0 char)
//
// returns the encoded buffer. Must be free'd by caller. Length is given by
//	outputLength.
//
char *RK_NewBase64Encode(
                         const void *buffer,
                         size_t length,
                         bool separateLines,
                         size_t *outputLength)
{
    const unsigned char *inputBuffer = (const unsigned char *)buffer;
    
#define MAX_NUM_PADDING_CHARS 2
#define OUTPUT_LINE_LENGTH 64
#define INPUT_LINE_LENGTH ((OUTPUT_LINE_LENGTH / BASE64_UNIT_SIZE) * BINARY_UNIT_SIZE)
#define CR_LF_SIZE 2
    
    //
    // Byte accurate calculation of final buffer size
    //
    size_t outputBufferSize =
    ((length / BINARY_UNIT_SIZE)
     + ((length % BINARY_UNIT_SIZE) ? 1 : 0))
    * BASE64_UNIT_SIZE;
    if (separateLines) {
        outputBufferSize +=
        (outputBufferSize / OUTPUT_LINE_LENGTH) * CR_LF_SIZE;
    }
    
    //
    // Include space for a terminating zero
    //
    outputBufferSize += 1;
    
    //
    // Allocate the output buffer
    //
    char *outputBuffer = (char *)malloc(outputBufferSize);
    if (!outputBuffer) {
        return NULL;
    }
    
    size_t i = 0;
    size_t j = 0;
    const size_t lineLength = separateLines ? INPUT_LINE_LENGTH : length;
    size_t lineEnd = lineLength;
    
    while (true)
    {
        if (lineEnd > length) {
            lineEnd = length;
        }
        
        for (; i + BINARY_UNIT_SIZE - 1 < lineEnd; i += BINARY_UNIT_SIZE) {
            //
            // Inner loop: turn 48 bytes into 64 base64 characters
            //
            outputBuffer[j++] = (char)base64EncodeLookup[(inputBuffer[i] & 0xFC) >> 2];
            outputBuffer[j++] = (char)base64EncodeLookup[((inputBuffer[i] & 0x03) << 4)
                                                         | ((inputBuffer[i + 1] & 0xF0) >> 4)];
            outputBuffer[j++] = (char)base64EncodeLookup[((inputBuffer[i + 1] & 0x0F) << 2)
                                                         | ((inputBuffer[i + 2] & 0xC0) >> 6)];
            outputBuffer[j++] = (char)base64EncodeLookup[inputBuffer[i + 2] & 0x3F];
        }
        
        if (lineEnd == length) {
            break;
        }
        
        //
        // Add the newline
        //
        outputBuffer[j++] = '\r';
        outputBuffer[j++] = '\n';
        lineEnd += lineLength;
    }
    
    if (i + 1 < length) {
        //
        // Handle the single '=' case
        //
        outputBuffer[j++] = (char)base64EncodeLookup[(inputBuffer[i] & 0xFC) >> 2];
        outputBuffer[j++] = (char)base64EncodeLookup[((inputBuffer[i] & 0x03) << 4)
                                                     | ((inputBuffer[i + 1] & 0xF0) >> 4)];
        outputBuffer[j++] = (char)base64EncodeLookup[(inputBuffer[i + 1] & 0x0F) << 2];
        outputBuffer[j++] =	'=';
    }
    else if (i < length) {
        //
        // Handle the double '=' case
        //
        outputBuffer[j++] = (char)base64EncodeLookup[(inputBuffer[i] & 0xFC) >> 2];
        outputBuffer[j++] = (char)base64EncodeLookup[(inputBuffer[i] & 0x03) << 4];
        outputBuffer[j++] = '=';
        outputBuffer[j++] = '=';
    }
    outputBuffer[j] = 0;
    
    //
    // Set the output length and return the buffer
    //
    if (outputLength) {
        *outputLength = j;
    }
    return outputBuffer;
}

@implementation Base64
//
// dataFromBase64String:
//
// Creates an NSData object containing the base64 decoded representation of
// the base64 string 'aString'
//
// Parameters:
//    aString - the base64 string to decode
//
// returns the autoreleased NSData representation of the base64 string
//
+ (NSData *)rk_dataFromBase64String:(NSString *)aString
{
    NSData *data = [aString dataUsingEncoding:NSASCIIStringEncoding];
    size_t outputLength;
    void *outputBuffer = RK_NewBase64Decode([data bytes], [data length], &outputLength);
    NSData *result = [NSData dataWithBytes:outputBuffer length:outputLength];
    free(outputBuffer);
    return result;
}

//
// base64EncodedString
//
// Creates an NSString object that contains the base 64 encoding of the
// receiver's data. Lines are broken at 64 characters long.
//
// returns an autoreleased NSString being the base 64 representation of the
//	receiver.
//
+ (NSString *)rk_base64EncodedString:(NSData *)aData
{
    size_t outputLength = 0;
    char *outputBuffer = RK_NewBase64Encode([aData bytes], [aData length], false, &outputLength);
    
    NSString *result =
    [[NSString alloc]
     initWithBytes:outputBuffer
     length:outputLength
     encoding:NSASCIIStringEncoding];
    free(outputBuffer);
    return result;
}
@end
